File: v1/modules/signature/signature.js
require('dotenv').config();
require('../components/enc-base64-min.js')();
const CryptoJS = require('../rollups/hmac-sha256.js');

let accessKey = process.env['NCR_ACCESS_KEY'];

let makeSignature = (method, path, timestamp)=>{
	let secretKey = process.env['NCR_SECRET_KEY'];
	let hmac = CryptoJS.algo.HMAC.create(CryptoJS.algo.SHA256, secretKey);
  let sig = `${method} ${path}\n${timestamp}\n${accessKey}`;
  hmac.update(sig);
	let hash = hmac.finalize();
	return hash.toString(CryptoJS.enc.Base64);
}

let getHeader = (method,path)=>{
  let timestamp = String(new Date().valueOf());
  let sig = makeSignature(method,path,timestamp);
  return {
    'x-ncp-apigw-timestamp':timestamp,
    'x-ncp-iam-access-key':accessKey,
    'x-ncp-apigw-signature-v2':sig
  }
}

module.exports = getHeader;

File: v2/signature/signature.go
package signature

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"os"
	"strconv"
	"time"

	"github.com/joho/godotenv"
)

func init() {
	if _, err := os.Stat(".env"); err == nil {
		err = godotenv.Load()
		if err != nil {
			os.Stderr.WriteString(err.Error())
		}
	}
}

func makeSignature(method *string, path *string, timestamp *string) string {
	accessKey := os.Getenv("NCR_ACCESS_KEY")
	secretKey := os.Getenv("NCR_SECRET_KEY")
	digest := hmac.New(sha256.New, []byte(secretKey))
	sig := *method + " " + *path + "\n" + *timestamp + "\n" + accessKey
	digest.Write([]byte(sig))
	return base64.StdEncoding.EncodeToString(digest.Sum(nil))
}

func GetHeader(method *string, path *string) *map[string]string {
	timestamp := strconv.FormatInt(time.Now().UnixNano()/int64(time.Millisecond), 10)
	accessKey := os.Getenv("NCR_ACCESS_KEY")
	sig := makeSignature(method, path, &timestamp)
	headers := make(map[string]string)
	headers["x-ncp-apigw-timestamp"] = timestamp
	headers["x-ncp-iam-access-key"] = accessKey
	headers["x-ncp-apigw-signature-v2"] = sig
	return &headers
}
